### DISCUSSION TITLE
The future of Payload localization and routing

### DISCUSSION THREAD

**Discussion by HarleySalas** [payloadcms/payload#4004] [Feature Requests & Ideas]
I understand that many plugins, such-as `@payloadcms/plugin-nested-docs` will be improved and addressed in the near future.

However, I wanted to take the time to chime in and mention some of the things that I think should be considered when moving forward with nested documents and/or pages in general.

A fully localized application (for good SEO and user experience) means translating **all** slugs and pathname urls, completely.

Currently, I understand the structure of [/payload/templates/website](https://github.com/payloadcms/payload/tree/main/templates/website), however, you can see that some of those paths, such-as `posts` or `projects` are hardcoded and in a scenario where you're trying to build a fully localized website, that isn't acceptable.

In my websites right now, I am using hooks to generate a `slug`, as well as a full `pathname` for each `Page` that I create. Another problem with this, depending on how you do routing for your localized pathnames, is that a user might receive a link to your website with `/sp/spanish/pathname`, but their language is English and they wish to visit the English version, so they may modify the path before even visiting your site, to `/en/spanish/pathname` and with that localized pathname, at least for the way I'm doing it now would create an issue. I understand there are *ways* to resolve this. However, I think Payload could do many things to make the process of routing and localization easy for any scenario.

Especially since the future of Payload is leaning in to NextJS, I really think you should aim to achieve some sort of feature parity with the next app router. What about taking advantage of nested layouts? Unless we hardcode things, I'm not sure how we can really take advantage of those sort of things.

What if we could have a setup similar to nested docs now, where blog posts would be under a separate posts collection, but we could create a `template` page in our Pages collection? I think there are many scenarios where the  data structure is consistent, but the page design, or components would want to be changed. I can see myself wanting to have a `Products` collection for example, where I record all of the data about my product, but a product *template* in my pages, where I would then be able to plug that data in to different fields.

I understand the level of complexity is quite a bit, but if you can accomplish:

- Simple integration with all, or most NextJS App router features,
- **Full** localization of all pathnames
- Page templates, with the ability to source data from fields on other collections

I genuinely believe it would be a major factor towards substantial growth of Payload. If those things can't all be achieved via the payload core package, or an official plugin, I think there should at least be an effort to publish more documentation, or learning materials for accomplishing these things.

I'd love to hear some other thoughts as well, or even perhaps how some of the issues I brought up can be resolved. It'd be really nice to know what the position of the Payload team is on these topics, to know if these sort of things are on the map and how much of a priority they may, or may not have.

**Comment by jancbeck** [comment]
You may want to check out [next-intl](https://next-intl-docs.vercel.app) for that. It supports locale detection and redirects. You can generate localized pathnames at build time or dynamically.

**Reply by HarleySalas** [reply]
I use `next-intl`. Since writing this, my approach is that I need to make a hidden json admin field, which stores all pathnames for each locale in one place, so that `next-intl` can create the alternate urls. I still wish there was some more thought put in to things like this, because really, I find that a full-featured i18n library is not really necessary for most projects. If payload expanded on it's functionality just a bit, you could just use `next-i18n-router` and for 99% of scenarios be totally fine without an i18n library/extra dependencies.

**Comment by amannn** [comment]
Hey, maintainer of `next-intl` here!

It's true that `next-intl` provides localized pathnames, but for pages that represent dynamic content from a CMS, it's not what I'd suggest.

Instead, I can recommend storing a `locale` field on collections where you need document-level localization and then link them together with a `translations` relationship. Then, you can render them with a dynamic `slug` like `/stories/[slug]` and `next-intl` will handle the locale prefixing for you. 



The hardest part here IMO is implementing the locale switcher. It's tricky, because on most pages you only need to switch the locale (i.e. the pathname prefix), but for the pages with localized slugs you need to switch the slug as well.

I managed to get that working with a `@locales` parallel route:

```
src/app/[locale]/@locales/
  ...
  stories/[slug]/page.tsx
```


If somebody is interested in the code (and conceptual background), I've made this available as part of 🌐 [learn.next-intl.dev](https://learn.next-intl.dev#project-code) (see `app-router-cms-payload` a bit below). It also includes the typical production optimizations like a localized sitemap with alternate links / hreflang.

**Reply by mikkelwf** [reply]
I just implemented dynamic routes with payload and next-intl last week.

I made a virtual url-field (per collection and utilized the getPathname method to correctly translate the dynamic paths to each locale.

Supporting both locale='all' and the specific locale takes a bit of work, but it makes it possible to do a subsequent query on the resolved page like `payload.findById({collection: 'page', id: 123, select: {slug: true, url: true}, locale: 'all'})`, then you get all localized urls..
